tallycash/extension
# 🚓Nonce zwei polizei: Register internally when a tx is dropped

## Blast from the past 
<details>

  <summary>Mo-Do - Eins, Zwei, Polizei</summary>

  https://www.youtube.com/watch?v=dSy2DcATYUo&ab_channel=ExpandedMusic
  
> "Eins, Zwei, Polizei" ([English](https://en.wikipedia.org/wiki/English_language): One, Two, Police) is a 1994 song recorded by Italian [dance](https://en.wikipedia.org/wiki/Electronic_dance_music) act [Mo-Do](https://en.wikipedia.org/wiki/Mo-Do) and was released as his first [single](https://en.wikipedia.org/wiki/Single_(music)) from their 1995 album, Was Ist Das?

https://en.wikipedia.org/wiki/Eins,_Zwei,_Polizei
</details>


## What 

This PR registers in our internal bookkeeping if a tx is dropped in the mempool. 

## Why

Before this change, the assumption was that after a tx is broadcasted, it will have a corresponding event when it's either dropped or mined. When a tx is dropped, it does not generate an event. More precisely [our listener is not triggered](https://github.com/tallycash/extension/blob/30e7e760a24f341b9d4e0e557c9e68f79d8eece8/background/services/chain/index.ts#L1702)

This behavior can be observed [in the alchemy mempool watcher](https://dashboard.alchemy.com/mempool?app_id=7lfjs0o15s01hwie&search_input=&status=0&time_min=1668955342000&time_range_preset=lastDay) if the error is reasonable — skipping only 1 nonce, 574 after 572. If the error in nonce is more significant — 100+ — then the tx doesn't show up in the mempool watcher.

I couldn't find any official docs describing the exact rules here.

## How 

This PR introduces a code path that checks if a tx was dropped from mempool and if so cleans up our retrieve queue. 

⚠️ Before this change, we had another queue mechanism that flattened the retrievals of txs. It was removed in this PR. I couldn't find an implementation that felt to be a good balance of added complexity and added functionality. For context see #2102 

## Test 

⚠️Note for reviewers: This change is based on observations about how things work, and I don't have any docs to back-this up — although this behavior makes sense. Please be vigilant in testing and reviewing. 

- import new seed phrase and wait ~5 mins for the initial tx retrievals to settle
- test normal behavior
  - Apply [wrong-nonce.patch.txt](https://github.com/tallycash/extension/files/10049533/wrong-nonce.patch.txt)
or add a [wrong nonce above this line `nonce: <wrong_nonce>`](https://github.com/tallycash/extension/blob/81cffeea62af71ce38d3a25bcb5e92613ed6c693/ui/pages/Send.tsx#L139) 
    - if you choose a wrong nonce with only 1 skipped the tx will show up [in the alchemy mempool watcher](https://dashboard.alchemy.com/mempool?app_id=7lfjs0o15s01hwie&search_input=&status=0&time_min=1668955342000&time_range_preset=lastDay)
    - if you use nonce that's 100+ away from the current one on chain it won't show 
  - send assets on mainnet or polygon > See the pending tx, that does not appear on etherscan
  - After 5-10 minutes the pending tx should be marked as dropped
  - you should see a warning in the console `Tx has 0x... is found in our local registry but not on chain`
- test cleanup after restart
  - Apply [wrong-nonce.patch.txt](https://github.com/tallycash/extension/files/10049533/wrong-nonce.patch.txt)
or add a [wrong nonce above this line `nonce: <wrong_nonce>`](https://github.com/tallycash/extension/blob/81cffeea62af71ce38d3a25bcb5e92613ed6c693/ui/pages/Send.tsx#L139) 
  - send assets on mainnet or polygon > See the pending tx, that does not appear on etherscan
  - restart the extension 
  - wait 5-10 minutes
  - the tx should be marked as dropped
  - you should see a warning in the console `Tx has 0x... is found in our local registry but not on chain`

Latest build: [extension-builds-2527](https://github.com/tallycash/extension/suites/9732265074/artifacts/466925410) (as of Wed, 07 Dec 2022 20:57:11 GMT).
